README.md
# distributed-process-cluster

Akka inspired cluster membership to keep track of what nodes are part of a the cluster.


ChangeLog.md
# Changelog for distributed-process-cluster

## Unreleased changes


test/Spec.hs
import Test.Framework
import Test.Framework.Providers.HUnit
import Test.HUnit
import Control.Distributed.Process.Cluster.Internal.VClock as VC

main :: IO ()
main = defaultMain [ testCase "insert" test_insert
                   , testCase "lookup" test_lookup
                   , testCase "merge" test_merge
                   , testCase "inc" test_inc
                   , testCase "incDefault" test_incDefault
                   , testCase "relate_before 1" test_relate_before1
                   , testCase "relate_before 2" test_relate_before2
                   , testCase "relate_before 3" test_relate_before3
                   , testCase "relate_concurrent 1" test_relate_concurrent1
                   , testCase "relate_concurrent 2" test_relate_concurrent2]


test_insert :: Assertion
test_insert = do
    insert ("0", 1) c @?= fromList [("0", 1), ("a", 2)]
    insert ("a", 1) c @?= fromList [("a", 2)]
    insert ("a", 3) c @?= fromList [("a", 3)]
        where c = fromList [("a", 2)]

test_lookup = do
    VC.lookup "a" c @?= Just 1
    VC.lookup "a" (empty :: VClock String Int) @?= Nothing
        where c = insert ("a", 1) empty

test_merge = VC.merge c1 c2 @?= merged
    where c1 = fromList [("a", 1), ("b", 1)]
          c2 = fromList [("a", 1), ("b", 2), ("c", 3)]
          merged = fromList [("a", 1), ("b", 2), ("c", 3)]

test_inc = do
    VC.inc "a" c @?= fromList [("a", 2)]
    VC.inc "b" c @?= fromList [("a", 1)]
        where c = fromList [("a", 1)]


test_incDefault = do
    VC.incDefault "a" 0 c @?= fromList [("a", 2)]
    VC.incDefault "b" 2 c @?= fromList [("a", 1), ("b", 2)]
    VC.incDefault "0" 2 c @?= fromList [("0", 2), ("a", 1)]
        where c = fromList [("a", 1)]

test_relate_before1 = let c1 = insert ("a", 0) empty 
                          c2 = inc "a" c1
                       in c1 `relates` c2 @?= Before

test_relate_before2 = let clock1_1 = empty :: VClock String Int
                          clock2_1 = incDefault "1" 0 clock1_1
                          clock3_1 = incDefault "2" 0 clock2_1
                          clock4_1 = incDefault "1" 0 clock3_1
                          clock1_2 = empty :: VClock String Int
                          clock2_2 = incDefault "1" 0 clock1_2
                          clock3_2 = incDefault "2" 0 clock2_2
                          clock4_2 = incDefault "1" 0 clock3_2
                          clock5_2 = incDefault "3" 0 clock4_2
                       in clock4_1 `relates` clock5_2 @?= Before

test_relate_before3 = let clock1_1 = empty :: VClock String Int
                          clock2_1 = incDefault "2" 0 clock1_1
                          clock3_1 = incDefault "2" 0 clock2_1
                          clock1_2 = empty :: VClock String Int
                          clock2_2 = incDefault "1" 0 clock1_2
                          clock3_2 = incDefault "2" 0 clock2_2
                          clock4_2 = incDefault "2" 0 clock3_2
                          clock5_2 = incDefault "3" 0 clock4_2
                        in do
                            clock3_1 `relates` clock5_2 @?= Before
                            clock5_2 `relates` clock3_1 @?= After
                      
test_relate_concurrent1 = let clock1_1 = empty :: VClock String Int
                              clock2_1 = incDefault "1" 0 clock1_1
                              clock1_2 = empty :: VClock String Int
                              clock2_2 = incDefault "2" 0 clock1_2
                           in clock2_1 `relates` clock2_2 @?= Concurrent


test_relate_concurrent2 = let clock1 = fromList [("1", 0), ("2", 1), ("3", 0)]
                              clock2 = fromList [("1", 0), ("2", 3)]
                           in do
                               clock1 `relates` clock2 @?= Concurrent
                               clock2 `relates` clock1 @?= Concurrent


app/Main.hs
{-# LANGUAGE GeneralizedNewtypeDeriving, LambdaCase #-}
module Main where
import Control.Distributed.Process.Cluster
import Control.Monad
import Control.Concurrent (threadDelay)
import Network.Transport
import Network.Transport.TCP (createTransport, defaultTCPParameters)
import Control.Distributed.Process
import Control.Distributed.Process.Node
import Data.ByteString.Char8 as BS (pack,unpack,concat)
import System.Environment


makeNodeId :: String -> NodeId
makeNodeId port = NodeId . EndPointAddress . BS.concat $ [BS.pack ("localhost"), BS.pack (":" ++ port), BS.pack ":0"]

main = do
    port : seeds <- getArgs

    Right t <- createTransport "127.0.0.1" port (\s -> ("localhost", s)) defaultTCPParameters
    node <- newLocalNode t initRemoteTable

    runCluster node (map makeNodeId seeds) $ do
        self <- getSelfPid
        forever $ do
            liftIO $ threadDelay 5000000
            getpeers >>= say . show


src/Control/Distributed/Process/Gossip.hs
{-# LANGUAGE DeriveGeneric
            , LambdaCase
            , ImplicitParams  #-}
module Control.Distributed.Process.Gossip(runCluster, getpeers) where

import Data.Binary (Binary)
import GHC.Generics
import System.Random
import Data.List
import Control.Concurrent (threadDelay)
import Control.Concurrent.STM
import Control.Monad.Reader
import Control.Distributed.Process
import Control.Distributed.Process.Serializable
import Control.Distributed.Process.Node
import Control.Distributed.Process.Gossip.Internal.VClock as VClock

type Version = VClock ProcessId Integer

data Peer = Peer ProcessId PeerState
    deriving (Show, Generic)

data GetPeers = GetPeers
    deriving (Show, Generic)

data Protocol = Join ProcessId
              | JoinAck ProcessId Version [Peer]
              | Gossip ProcessId Version [Peer]
              | GossipTimer
              | Ping ProcessId | Pong
              deriving(Show, Eq, Generic)


data PeerState = Up | Down deriving (Show, Eq, Generic)

instance Binary Protocol where
instance Binary PeerState where
instance Binary Peer where
instance Binary GetPeers

instance Eq Peer where
    (Peer p1 _) == (Peer p2 _) = p1 == p2

data ClusterState  = ClusterState {
    _version :: Version,
    _peers :: [Peer]
}

data Config = Config { state :: !(TVar ClusterState) }

put :: (MonadIO m) => (?config :: Config) => ClusterState -> m ClusterState
put cs = liftIO $ atomically $ swapTVar (state ?config) cs

get :: (MonadIO m) => (?config :: Config) => m ClusterState
get = liftIO $ readTVarIO (state ?config)

modify :: (MonadIO m) => (?config :: Config) => (ClusterState -> ClusterState) -> m ()
modify f = liftIO $ atomically $ modifyTVar' (state ?config) f

sample :: [g] -> IO g
sample lst = do
    i <- randomRIO (0, length lst -1)
    return $ lst !! i

handle GossipTimer = do
    self <- getSelfPid
    ClusterState version peers <- get
    Peer p _ <- liftIO $ sample (filter (\(Peer p s) -> s == Up) peers)
    send p (Gossip self version peers)

handle (Gossip pid v1 peers) = do
    ClusterState v2 peers2 <- get
    put $ case v1 `relates` v2 of
        Same -> ClusterState v1 peers
        Before -> ClusterState v2 peers2
        After -> ClusterState v1 peers
        Concurrent -> ClusterState (v1 `merge` v2) (peers `union` peers2)

    ClusterState _ npeers <- get
    return ()

handle (Join id) = do
    self <- getSelfPid
    modify $ \s -> s { _version = VClock.inc self (_version s),
                       _peers = [Peer id Up] `union` (_peers s) }
    ClusterState v p <- get
    send id (JoinAck self v p)
    return ()

handle (Ping p) = do
    send p Pong

handlePeerRequest :: (?config :: Config) => SendPort [ProcessId] -> Process ()
handlePeerRequest sender = do
    ClusterState _ npeers <- get
    sendChan sender $  map (\(Peer p _) -> p) $ filter (\(Peer _ s) -> s == Up) npeers

getpeers :: Process [NodeId]
getpeers = do
    (sp, rp) <- newChan
    nsend clusterController (sp :: SendPort [ProcessId])
    receiveChan rp >>= return . map processNodeId


heartbeat :: (?config :: Config) => Process ()
heartbeat = forever $ do
    self <- getSelfPid
    liftIO $ threadDelay 5000000
    ClusterState _ peers <- get
    Peer p _ <- liftIO $ sample (filter (\(Peer p s) -> s == Up) peers)
    send p (Ping self)
    expectTimeout 1000000 >>= \case
        (Just Pong) -> return ()
        Nothing     -> modify $ \s -> s { _version = VClock.inc self (_version s),
                                          _peers = [Peer p Down] `union` (_peers s)}


clusterController = "Cluster:Controller"

gossip :: (?config :: Config) => ProcessId -> Process ()
gossip pid = do
    forever $ do
        liftIO $ threadDelay 5000000
        send pid GossipTimer


initialized  :: (?config :: Config) => Process ()
initialized = do
    pid <- getSelfPid
    say $ "Cluster started"

    spawnLocal $ gossip pid
    spawnLocal heartbeat

    forever $ receiveWait [ match handle,
                            match handlePeerRequest ]

uninitialized :: (?config :: Config) => [NodeId] -> Process ()
uninitialized [] = do
    pid <- getSelfPid

    modify $ \s -> s { _version = VClock.incDefault pid 0 (_version s),
                       _peers = [Peer pid Up] }
    initialized

uninitialized seeds = do
    pid <- getSelfPid
    seedNode <- liftIO $ sample seeds

    say $ "Trying to join " ++ show seedNode
    nsendRemote seedNode clusterController (Join pid)
    receiveWait [match $ joining]

joining (JoinAck rpid version peers) = do
    pid <- getSelfPid

    modify $ \s -> s { _version = VClock.incDefault pid 0 version, 
                       _peers = peers }

    say $ "Joined"
    initialized

runCluster :: LocalNode -> [NodeId] -> Process () -> IO ()
runCluster node seeds proc = do
    cstate <- newTVarIO initialState
    let ?config = Config cstate
    _ <- forkProcess node $ do
        getSelfPid >>= register clusterController
        uninitialized seeds
    runProcess node proc
    where initialState = ClusterState VClock.empty [] 
    

src/Control/Distributed/Process/Cluster/Internal/VClock.hs
module Control.Distributed.Process.Cluster.Internal.VClock (VClock
              , Relation(..)
              , empty
              , insert
              , fromList
              , toList
              , lookup
              , inc
              , incDefault
              , merge
              , relates
              ) where

import Data.Foldable (Foldable(foldMap), foldl')
import Prelude hiding (lookup)
import Data.Binary
import qualified Prelude as P (lookup)

newtype VClock a b = VClock [(a, b)]
    deriving (Show, Eq)

instance (Ord a, Ord b, Binary a, Binary b) => Binary (VClock a b) where
    put = put . toList
    get = get >>= \lst -> return $ fromList lst

data Relation = Same | Before | After | Concurrent
    deriving (Show,Eq)

instance Foldable (VClock a) where
    foldMap f (VClock vc) = foldMap f (map snd vc)

empty :: (Ord a) => VClock a b
empty = VClock []

insert :: (Ord a, Ord b) => (a, b) -> VClock a b -> VClock a b
insert (k, v) (VClock vc) = VClock $ go vc
    where go [] = [(k,v)]
          go (vc'@(k',v') : vcs')
            | k' < k    = vc' : go vcs' 
            | k' == k   = (if v >= v' then (k, v) else (k, v')) : vcs'
            | otherwise = (k, v) : vc' : vcs'

fromList :: (Ord a, Ord b) => [(a, b)] -> VClock a b
fromList = foldl' (flip insert) empty

toList :: VClock a b -> [(a,b)]
toList (VClock d) = d

lookup :: (Ord a) => a -> VClock a b -> Maybe b
lookup key (VClock vc) = P.lookup key vc

inc :: (Ord a, Ord b, Num b) => a -> VClock a b -> VClock a b
inc k c = case lookup k c of
            Just v -> insert (k, v + 1) c
            Nothing -> c

incDefault :: (Ord a, Ord b, Num b) => a -> b -> VClock a b -> VClock a b
incDefault k d c = case lookup k c of
                     Just v -> inc k c
                     Nothing -> insert (k, d) c

merge :: (Ord a, Ord b) => VClock a b -> VClock a b -> VClock a b
merge (VClock vc1) other = foldl' (flip insert) other vc1 

relates :: (Ord a, Ord b) => VClock a b -> VClock a b -> Relation
relates (VClock c1) (VClock c2) = relates' c1 c2 Same
    where relates' [] [] s = s
          relates' [] c2s s = if s == After then Concurrent else Before
          relates' c1s [] s = if s == Before then Concurrent else After
          relates' ((k1,v1):c1s) ((k2,v2):c2s) s = 
              case compare k1 k2 of
                EQ -> case compare v1 v2 of
                        EQ -> relates' c1s c2s s
                        LT -> if s == After  then Concurrent else relates' c1s c2s Before
                        _  -> if s == Before then Concurrent else relates' c1s c2s After
                LT -> case s of
                        Before -> Concurrent
                        _      -> relates' c1s ((k2,v2):c2s) After
                GT -> case s of
                        After -> Concurrent
                        _     -> relates' ((k1,v2):c1s) c2s Before
 


src/Control/Distributed/Process/Cluster.hs
{-# LANGUAGE DeriveGeneric
            , LambdaCase
            , ImplicitParams  #-}
module Control.Distributed.Process.Cluster(runCluster, getpeers) where

import Data.Binary (Binary)
import GHC.Generics
import System.Random
import Data.List
import Control.Concurrent (threadDelay)
import Control.Concurrent.STM
import Control.Monad.IO.Class
import Control.Monad
import Control.Distributed.Process
import Control.Distributed.Process.Serializable
import Control.Distributed.Process.Node
import Control.Distributed.Process.Cluster.Internal.VClock as VClock

type Version = VClock ProcessId Integer

data Peer = Peer ProcessId PeerState
    deriving (Show, Generic)

data GetPeers = GetPeers
    deriving (Show, Generic)

data Protocol = Join ProcessId
              | JoinAck ProcessId Version [Peer]
              | Gossip ProcessId Version [Peer]
              | GossipTimer
              | Ping ProcessId | Pong
              deriving(Show, Eq, Generic)


data PeerState = Up | Down deriving (Show, Eq, Generic)

instance Binary Protocol where
instance Binary PeerState where
instance Binary Peer where
instance Binary GetPeers

instance Eq Peer where
    (Peer p1 _) == (Peer p2 _) = p1 == p2

data ClusterState  = ClusterState {
    _version :: Version,
    _peers :: [Peer]
}

data Config = Config { state :: !(TVar ClusterState) }

put :: (MonadIO m) => (?config :: Config) => ClusterState -> m ClusterState
put cs = liftIO $ atomically $ swapTVar (state ?config) cs

get :: (MonadIO m) => (?config :: Config) => m ClusterState
get = liftIO $ readTVarIO (state ?config)

modify :: (MonadIO m) => (?config :: Config) => (ClusterState -> ClusterState) -> m ()
modify f = liftIO $ atomically $ modifyTVar' (state ?config) f

sample :: [g] -> IO g
sample lst = do
    i <- randomRIO (0, length lst -1)
    return $ lst !! i

handle GossipTimer = do
    self <- getSelfPid
    ClusterState version peers <- get
    Peer p _ <- liftIO $ sample (filter (\(Peer p s) -> s == Up) peers)
    send p (Gossip self version peers)

handle (Gossip pid v1 peers) = do
    ClusterState v2 peers2 <- get
    put $ case v1 `relates` v2 of
        Same -> ClusterState v1 peers
        Before -> ClusterState v2 peers2
        After -> ClusterState v1 peers
        Concurrent -> ClusterState (v1 `merge` v2) (peers `union` peers2)

    ClusterState _ npeers <- get
    return ()

handle (Join id) = do
    self <- getSelfPid
    modify $ \s -> s { _version = VClock.inc self (_version s),
                       _peers = [Peer id Up] `union` (_peers s) }
    ClusterState v p <- get
    send id (JoinAck self v p)
    return ()

handle (Ping p) = do
    send p Pong

handlePeerRequest :: (?config :: Config) => SendPort [ProcessId] -> Process ()
handlePeerRequest sender = do
    ClusterState _ npeers <- get
    sendChan sender $  map (\(Peer p _) -> p) $ filter (\(Peer _ s) -> s == Up) npeers

getpeers :: Process [NodeId]
getpeers = do
    (sp, rp) <- newChan
    nsend clusterController (sp :: SendPort [ProcessId])
    receiveChan rp >>= return . map processNodeId


heartbeat :: (?config :: Config) => Process ()
heartbeat = forever $ do
    self <- getSelfPid
    liftIO $ threadDelay 5000000
    ClusterState _ peers <- get
    Peer p _ <- liftIO $ sample (filter (\(Peer p s) -> s == Up) peers)
    send p (Ping self)
    expectTimeout 1000000 >>= \case
        (Just Pong) -> return ()
        Nothing     -> modify $ \s -> s { _version = VClock.inc self (_version s),
                                          _peers = [Peer p Down] `union` (_peers s)}


clusterController = "Cluster:Controller"

gossip :: (?config :: Config) => ProcessId -> Process ()
gossip pid = do
    forever $ do
        liftIO $ threadDelay 5000000
        send pid GossipTimer


initialized  :: (?config :: Config) => Process ()
initialized = do
    pid <- getSelfPid
    say $ "Cluster started"

    spawnLocal $ gossip pid
    spawnLocal heartbeat

    forever $ receiveWait [ match handle,
                            match handlePeerRequest ]

uninitialized :: (?config :: Config) => [NodeId] -> Process ()
uninitialized [] = do
    pid <- getSelfPid

    modify $ \s -> s { _version = VClock.incDefault pid 0 (_version s),
                       _peers = [Peer pid Up] }
    initialized

uninitialized seeds = do
    pid <- getSelfPid
    seedNode <- liftIO $ sample seeds

    say $ "Trying to join " ++ show seedNode
    nsendRemote seedNode clusterController (Join pid)
    receiveWait [match $ joining]

joining (JoinAck rpid version peers) = do
    pid <- getSelfPid

    modify $ \s -> s { _version = VClock.incDefault pid 0 version, 
                       _peers = peers }

    say $ "Joined"
    initialized

runCluster :: LocalNode -> [NodeId] -> Process () -> IO ()
runCluster node seeds proc = do
    cstate <- newTVarIO initialState
    let ?config = Config cstate
    _ <- forkProcess node $ do
        getSelfPid >>= register clusterController
        uninitialized seeds
    runProcess node proc
    where initialState = ClusterState VClock.empty [] 
    
